pubnub-core/src/subscription/registry.rs

```diff
@@ -0,0 +1,72 @@
+use super::mvec::MVec;
+use std::collections::{hash_map::Entry, HashMap};
+
+/// A registry of channels.
+#[derive(Debug)]
+pub(crate) struct Registry<T> {
+    pub(super) map: HashMap<String, MVec<T>>,
+}
+
+/// Newtype to protect access to the registry ID.
+#[derive(Debug, Copy, Clone, PartialOrd, Ord, PartialEq, Eq)]
+pub struct ID(usize);
+
+#[derive(Debug)]
+pub enum RegistrationEffect {
+    NewName,
+    ExistingName,
+}
+
+#[derive(Debug)]
+pub enum UnregistrationEffect {
+    NameErazed,
+    NamePreserved,
+}
+
+impl<T> Registry<T> {
+    pub fn new() -> Self {
+        Self {
+            map: HashMap::new(),
+        }
+    }
+
+    pub fn register<S: Into<String>>(&mut self, name: S, value: T) -> (ID, RegistrationEffect) {
+        let entry = self.map.entry(name.into());
+
+        let effect = match &entry {
+            Entry::Vacant(_) => RegistrationEffect::NewName,
+            Entry::Occupied(_) => RegistrationEffect::ExistingName,
+        };
+
+        let mvec = entry.or_default();
+        let id = mvec.counter();
+        mvec.push(value);
+
+        (ID(id), effect)
+    }
+
+    pub fn unregister(&mut self, name: &str, id: ID) -> Option<(T, UnregistrationEffect)> {
+        let ID(id) = id;
+
+        let mvec = self.map.get_mut(name)?;
+        let removed = mvec.remove(id)?;
+
+        let effect = if mvec.is_empty() {
+            self.map.remove(name);
+            UnregistrationEffect::NameErazed
+        } else {
+            UnregistrationEffect::NamePreserved
+        };
+
+        Some((removed, effect))
+    }
+
+    // TODO: provide better interface for iteration over mutable items.
```

Comment:
Any specific ideas in mind?

Reply:
There are other todos at use locations that provide more hints. In essence, I'd replace this with just an iterator over the `MVec` values.

Reply:
Not entirely sure I follow. The only use of the iterator is accessed through a `HashMap<MVec<T>>`, so the call site needs to gain access to it through the `HashMap` interface. Is the idea here to combine the two interfaces?

Reply:
Pretty much, yes. We don't really need to expose the internal `MVec<T>` in the interface, all we need is to be able to enumerate the listeners for a certain route:

https://github.com/pubnub/rust/blob/dd3729ef2b9737b6c7ba008bd31415900e743ef1/pubnub-core/src/subscription/subscribe_loop.rs#L180-L182

This is the only call-site we currently have.